src/providers/VisibleTournamentsProvider.js

```diff
@@ -0,0 +1,68 @@
+import {useMemo, useRef, useState} from 'react';
+
+import VisibleTournamentsContext from '../contexts/VisibleTournamentsContext';
+import VisibleTournamentsTrackingContext from '../contexts/VisibleTournamentsTrackingContext';
+
+import useDebounce from '../use/useDebounce';
+
+const VISIBILITY_CHANGE_DEBOUNCE_WAIT_MS = 250;
+
+const VisibleTournamentsProvider = ({ children }) => {
+    const [visibleTournaments, setVisibleTournaments] = useState(new Set());
+
+    const watchedElementToTournamentRef = useRef(new Map());
+
+    const [setVisibleTournamentsDebounced] = useDebounce(setVisibleTournaments, VISIBILITY_CHANGE_DEBOUNCE_WAIT_MS);
+
+    const intersectionObserverRef = useRef(
+        new IntersectionObserver((entries) => {
+            setVisibleTournamentsDebounced((visibleTournaments) => {
+                const currentlyVisibleTournaments = new Set(visibleTournaments);
+
+                entries.forEach((entry) => {
+                    const props = watchedElementToTournamentRef.current.get(entry.target);
+
+                    // Shouldn't happen but what do I know.
+                    if (!props) {
```

Comment:
If this shouldn't happen but we want to be defensive (which I approve of!), maybe throw or at least log an error if it _does_ happen to make debugging easier?

Reply:
Done.